### PR TITLE
Implement add

### PR DESCRIPTION
**Pull request by Bihaqo**

### CODE REVIEW

File: ttax/decompositions.py [hunk 1]
```diff
@@ -4,8 +4,10 @@
 
 from ttax.base_class import TT
 from ttax.base_class import TTMatrix
+from ttax.ops import tt_vmap
 
 
+@tt_vmap()
```
Comment (Bihaqo) [inline ttax/decompositions.py]: Should be in the other pull request

File: ttax/ops.py [hunk 2]
```diff
@@ -8,20 +8,42 @@
...
-    if tt_arg.num_batch_dims == 0:
-      return func(*args, **kwargs)
-    else:
-      # Vmap everything num_batch_dims times.
-      vmapped = func
-      for _ in range(tt_arg.num_batch_dims):
-        vmapped = jax.vmap(vmapped)
-      return vmapped(*args, **kwargs)
-  return vectorized_func
+def tt_vmap(num_batch_args=None):
```
Comment (Bihaqo) [inline ttax/ops.py]: Should be in the other pull request

File: ttax/base_class.py [hunk 3]
```diff
@@ -52,6 +61,49 @@ def num_batch_dims(self):
...
+                                   remainder, sliced_core)
+          remainder = None
+        new_tt_cores.append(sliced_core)
+
+    if remainder is not None:
+      # The reminder obtained from collapsing the last cores.
+      new_tt_cores[-1] = jnp.einsum('...aib,...bd->...aid', 
+                                    new_tt_cores[-1], remainder)
+      remainder = None
+    # TODO: infer the output ranks and shape.
```
Comment (Bihaqo) [inline ttax/base_class.py]: Remove

File: ttax/base_class.py [hunk 4]
```diff
@@ -76,3 +128,59 @@ def num_batch_dims(self):
   @property
   def is_tt_matrix(self):
     return True
+  
+  def __getitem__(self, slice_spec):
```
Comment (Bihaqo) [inline ttax/base_class.py:131]: Add docstring

File: ttax/base_class.py [hunk 5]
```diff
@@ -76,3 +128,59 @@ def num_batch_dims(self):
...
+
+    n = self.ndim
```
Comment (Bihaqo) [inline ttax/base_class.py]: n is usually used for the size of one mode. Lets either use d, ndim, or self.ndim (without creating a separate variable)

File: ttax/base_class.py [hunk 6]
```diff
@@ -76,3 +128,59 @@ def num_batch_dims(self):
...
+    if len(slice_spec) != 2 * n:
+      raise ValueError('Expected %d indices, got %d' % (2 * n, len(slice_spec)))
+    for i in range(n):
+      if isinstance(slice_spec[i], slice) != isinstance(slice_spec[n+i], slice):
+        raise ValueError('Elements i_%d and j_%d should be the same type.' % (i,
```
Comment (Bihaqo) [inline ttax/base_class.py]: Elements i_%d and j_%d should be of the same type, instead: %s and %s.' % (i, i, slice_spec[i], slice_spec[n+i])

File: ttax/base_class.py [hunk 7]
```diff
@@ -76,3 +128,59 @@ def num_batch_dims(self):
...
+      new_tt_cores[-1] = jnp.einsum('...aijb,...bd->...aijd', 
+                                    new_tt_cores[-1], remainder)
+      remainder = None
+
+    return TTMatrix(new_tt_cores)
+
+
+class BatchIndexing:
+  def __init__(self, tt):
+    self.tt = tt
```
Comment (Bihaqo) [inline ttax/base_class.py]: Better to store it as self._tt -- since you don't expect anyone needing access it, it's better to use the "protected naming"

File: ttax/ops.py [hunk 8]
```diff
@@ -108,3 +130,138 @@ def matmul(a, b):
   )
   func = to_function(tt_einsum)
   return func(a, b)
+
+
+def add(tt_a, tt_b):
+  """Returns a TensorTrain corresponding to elementwise sum tt_a + tt_b.
+  The shapes of tt_a and tt_b should coincide.
+  Supports broadcasting:
```
Comment (Bihaqo) [inline ttax/ops.py]: Supports broadcasting, e.g. you can add a tensor train with batch size 7 and a tensor train with batch size 1:
  tt_batch.add(tt_single.batch_loc[np.newaxis])
where tt_single.batch_loc[np.newaxis] creates a singleton batch dimension.

Comment (Bihaqo) [reply ttax/ops.py]: FYI Dim: np.newaxis is just None, but looks a bit more readable.

File: ttax/ops.py [hunk 9]
```diff
@@ -108,3 +130,138 @@ def matmul(a, b):
...
+    tt_b: TT or TT-Matrix
+  Returns
+    TT or TT-Matrix
+  Raises
+    ValueError if the arguments shapes do not coincide.
+  """
+  if not are_shapes_equal(tt_a, tt_b):
+    raise ValueError('Types of the arguments or their tensor '
+                     'shapes are different, addition is not '
+                     ' available.')
```
Comment (Bihaqo) [inline ttax/ops.py]: No need to use space in the begining here

File: ttax/ops.py [hunk 10]
```diff
@@ -108,3 +130,138 @@ def matmul(a, b):
...
+  if not are_batches_broadcastable(tt_a, tt_b):
+    raise ValueError('The batch sizes are different and not 1, '
+                     'broadcasting is not available.')
+  # batches are not supported yet
```
Comment (Bihaqo) [inline ttax/ops.py]: remove

File: ttax/ops.py [hunk 11]
```diff
@@ -108,3 +130,138 @@ def matmul(a, b):
...
+
+
+def are_shapes_equal(tt_a, tt_b):
+  """Returns the result of equality check of 2 tensors' shapes: 
+  True if shapes are equal and False otherwise.
+  The arguments should be both TT-tensors or both TT-matrices.
+  The arguments should have the same tensor shape.
+  e.g. for TT core (*, *, ..., *, r_{i}, n_{i}, r_{i+1}):
+  axis *, *, ..., * correspond to the batch shape,
+  axis r_{i}, n_{i}, r_{i+1} correspond to the tensor shape.
```
Comment (Bihaqo) [inline ttax/ops.py]: I'm confused by this. Maybe just

The arguments should have the same tensor shape but potentially different TT-ranks

Instead of these four lines?

File: ttax/ops.py [hunk 12]
```diff
@@ -108,3 +130,138 @@ def matmul(a, b):
...
+  Args:
+    tt_a: TT or TT-Matrix
+    tt_b: TT or TT-Matrix
+  Returns:
+    tensor_check: bool
```
Comment (Bihaqo) [inline ttax/ops.py:264]: Doesn't matter here, but FYI: you can type annotate outputs like this:
def are_shapes_equal(tt_a, tt_b) -> bool:

File: ttax/ops.py [hunk 13]
```diff
@@ -108,3 +130,138 @@ def matmul(a, b):
...
+  """
+  tensor_check = True
+  if tt_a.is_tt_matrix != tt_b.is_tt_matrix:
+    tensor_check = False
+  if tt_a.raw_tensor_shape != tt_b.raw_tensor_shape:
```
Comment (Bihaqo) [inline ttax/ops.py]: You can't compare lists like this. Something like
np.any(np.array(tt_a.raw_tensor_shape) != np.array(tt_b.raw_tensor_shape))

File: ttax/ops.py [hunk 14]
```diff
@@ -108,3 +130,138 @@ def matmul(a, b):
...
+    tensor_check = False
+  return tensor_check
+
+
+def are_batches_broadcastable(tt_a, tt_b):
+  """Returns the result of compatibility check of 2 tensors' batches: 
+  True if batches are compatible and False otherwise.
+  The batch sizes should be equal otherwise at least one of them 
+  should equal to 1 for broadcasting to be available.
+  e.g. for TT core (*, *, ..., *, r_{i}, n_{i}, r_{i+1}):
```
Comment (Bihaqo) [inline ttax/ops.py]: remove this e.g., it's still confusing

File: ttax/ops.py [hunk 15]
```diff
@@ -108,3 +130,138 @@ def matmul(a, b):
...
+  axis r_{i}, n_{i}, r_{i+1} correspond to the tensor shape.
+  Args:
+    tt_a: TT or TT-Matrix
+    tt_b: TT or TT-Matrix
+  Returns:
+    batch_check: bool
+  """
+  batch_check = True
+  if tt_a.num_batch_dims != tt_b.num_batch_dims:
+    batch_check = False
```
Comment (Bihaqo) [inline ttax/ops.py]: easier to do return False here to avoid using else below (and thus avoid using one more level of identation)

File: ttax/base_class.py [hunk 16]
```diff
@@ -76,3 +127,60 @@ def num_batch_dims(self):
   @property
   def is_tt_matrix(self):
     return True
+  
+  def __getitem__(self, slice_spec):
+    """Basic indexing, returns a TTMatrix containing the specified region."""
```
Comment (Bihaqo) [inline ttax/base_class.py]: the specified element / slice

File: ttax/ops_test.py [hunk 17]
```diff
@@ -169,7 +198,41 @@ def testFlatInner(self):
...
+    right_shape = (4, 4, 4)
+    tt_a = random_.matrix(rng1, (left_shape, right_shape), tt_rank=3,
+                          dtype=dtype)
+    tt_b = random_.matrix(rng2, (left_shape, right_shape), tt_rank=[1, 4, 3, 1],
+                          dtype=dtype)
 
+    res_actual1 = ops.full(ops.add(tt_a, tt_b))
+    res_actual2 = ops.full(tt_a + tt_b)
+    res_desired = ops.full(tt_a) + ops.full(tt_b)
+    self.assertAllClose(res_actual1, res_desired, rtol=1e-3)
```
Comment (Bihaqo) [inline ttax/ops_test.py]: Does it not work with smaller rtol? E.g. 1e-5?

File: ttax/ops_test.py [hunk 18]
```diff
@@ -169,7 +198,41 @@ def testFlatInner(self):
...
+    self.assertAllClose(res_actual2, res_desired, rtol=1e-3)
+
+  def testAddBatch(self):
+    # Add two batches of TT-matrices.
+    rng1, rng2 = jax.random.split(jax.random.PRNGKey(0))
+    dtype = jnp.float32
+    left_shape = (2, 3, 4)
+    right_shape = (4, 4, 4)
+    tt_a = random_.matrix(rng1, (left_shape, right_shape), tt_rank=3,
```
Comment (Bihaqo) [inline ttax/ops_test.py:240]: Not a batch

File: ttax/ops_test.py [hunk 19]
```diff
@@ -77,6 +77,35 @@ def testFlatInner(self):
...
+    tt_b = random_.tensor(rng2, (2, 1, 3, 4), tt_rank=[1, 2, 4, 3, 1],
+                          dtype=dtype)
+
+    res_actual1 = ops.full(ops.add(tt_a, tt_b))
+    res_actual2 = ops.full(tt_a + tt_b)
+    res_desired = ops.full(tt_a) + ops.full(tt_b)
+    self.assertAllClose(res_actual1, res_desired)
+    self.assertAllClose(res_actual2, res_desired)
+
+  def testAddBatch(self):
```
Comment (Bihaqo) [inline ttax/ops_test.py]: Ideally I would also test broadcasting. You can use parametrized test for this
